staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/metrics_test.go

```diff
@@ -0,0 +1,268 @@
+			},
+			wantLabels: map[string]string{
+				"result":       "failure",
+				"failure_type": string(WebhookConversionInvalidConvertedObjectFailure),
+			},
+			expectedRequestValue: 1,
+			expectedLatencyCount: 4,
+		}, {
+			name: "test_conversion_failure_5",
+			fields: fields{
+				webhookConversionRequest: Metrics.webhookConversionRequest,
+				webhookConversionLatency: Metrics.webhookConversionLatency,
+			},
+			args: args{
+				ctx:       context.TODO(),
+				elapsed:   2 * time.Second,
+				errorType: WebhookConversionNoObjectsReturnedFailure,
+			},
+			wantLabels: map[string]string{
+				"result":       "failure",
+				"failure_type": string(WebhookConversionNoObjectsReturnedFailure),
+			},
+			expectedRequestValue: 1,
+			expectedLatencyCount: 5,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			m := &WebhookConversionMetrics{
+				webhookConversionRequest: tt.fields.webhookConversionRequest,
+				webhookConversionLatency: tt.fields.webhookConversionLatency,
+			}
+			m.ObserveWebhookConversionFailure(tt.args.ctx, tt.args.elapsed, tt.args.errorType)
+			expectCounterValue(t, "webhook_conversion_requests", tt.wantLabels, tt.expectedRequestValue)
+			expectHistogramCountTotal(t, "webhook_conversion_duration_seconds", tt.wantLabels, tt.expectedRequestValue)
+		})
+	}
+}
+
+func expectCounterValue(t *testing.T, name string, labelFilter map[string]string, wantCount int) {
+	metrics, err := legacyregistry.DefaultGatherer.Gather()
+	if err != nil {
+		t.Fatalf("Failed to gather metrics: %s", err)
+	}
+
+	counterSum := 0
+	for _, mf := range metrics {
+		if mf.GetName() != name {
+			continue // Ignore other metrics.
+		}
+		for _, metric := range mf.GetMetric() {
+			if !testutil.LabelsMatch(metric, labelFilter) {
+				continue
+			}
+			counterSum += int(metric.GetCounter().GetValue())
+		}
+	}
+	if wantCount != counterSum {
+		t.Errorf("Wanted count %d, got %d for metric %s with labels %#+v", wantCount, counterSum, name, labelFilter)
+		for _, mf := range metrics {
+			if mf.GetName() == name {
+				for _, metric := range mf.GetMetric() {
+					t.Logf("\tnear match: %s", metric.String())
+				}
+			}
+		}
+	}
+}
+
+func expectHistogramCountTotal(t *testing.T, name string, labelFilter map[string]string, wantCount int) {
+	metrics, err := legacyregistry.DefaultGatherer.Gather()
+	if err != nil {
+		t.Fatalf("Failed to gather metrics: %s", err)
+	}
+
+	counterSum := 0
+	for _, mf := range metrics {
+		if mf.GetName() != name {
+			continue // Ignore other metrics.
+		}
+		for _, metric := range mf.GetMetric() {
+			if !testutil.LabelsMatch(metric, labelFilter) {
+				continue
+			}
+			counterSum += int(metric.GetHistogram().GetSampleCount())
+		}
+	}
+	if wantCount != counterSum {
+		t.Errorf("Wanted count %d, got %d for metric %s with labels %#+v", wantCount, counterSum, name, labelFilter)
+		for _, mf := range metrics {
+			if mf.GetName() == name {
+				for _, metric := range mf.GetMetric() {
```

Comment:
I would move these into `component-base/metrics/testutils`, make them public and rename them `assertXCount` or whatnot.